{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.linear_model import LogisticRegressionCV\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.ensemble import VotingClassifier\n",
    "from sklearn.metrics import precision_recall_fscore_support\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_features = pd.read_csv('../dataset/train_original_with_id.csv')\n",
    "test_features = pd.read_csv('../dataset/test_original_with_id.csv')\n",
    "train_labels = pd.read_csv('../dataset/final_train.csv', usecols=[1])\n",
    "#train_index = train_labels[train_labels['Association'].apply(lambda x: x != 'F')].index\n",
    "test_labels = pd.read_csv('../dataset/final_test.csv', usecols=[1])\n",
    "#test_index = test_labels[test_labels['Association'].apply(lambda x: x != 'F')].index\n",
    "#train_features = train_features.iloc[train_index].values.tolist()\n",
    "#test_features = test_features.iloc[test_index].values.tolist()\n",
    "#train_features = train_features.values.tolist()\n",
    "#test_features = test_features.values.tolist()\n",
    "#train_labels =  list(train_labels.iloc[train_index]['Association'])\n",
    "#test_labels =  list(test_labels.iloc[test_index]['Association'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_features = train_features.iloc[:,:24]\n",
    "test_features = test_features.iloc[:,:24]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_labels = pd.read_csv('../dataset/final_train.csv',\n",
    "                          usecols=[1])\n",
    "train_labels['Association'] = train_labels['Association'].apply(lambda x: x != 'F')\n",
    "train_labels = list(train_labels['Association'])\n",
    "test_labels = pd.read_csv('../dataset/final_test.csv',\n",
    "                          usecols=[1])\n",
    "test_labels['Association'] = test_labels['Association'].apply(lambda x: x != 'F')\n",
    "test_labels = list(test_labels['Association'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.8379746835443038\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(array([0.58333333, 0.85444744]),\n",
       " array([0.20588235, 0.96941896]),\n",
       " array([0.30434783, 0.90830946]),\n",
       " array([ 68, 327]))"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf = LogisticRegression(solver='liblinear', multi_class='auto').fit(train_features, train_labels)\n",
    "print(sum([p == a for p, a in zip(clf.predict(test_features),test_labels)]) / len(test_labels))\n",
    "precision_recall_fscore_support(test_labels, clf.predict(test_features), average=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['shared_negation',\n",
       " 'PROPN',\n",
       " 'PUNCT',\n",
       " 'lrp_root',\n",
       " 'despite',\n",
       " 'polymorphism_root',\n",
       " 'formation',\n",
       " 'by',\n",
       " 'result',\n",
       " 'from',\n",
       " 'contribution',\n",
       " 'any_negation',\n",
       " 'DET',\n",
       " 'ADP',\n",
       " 'enos',\n",
       " 'genotype_root',\n",
       " 'homozygote',\n",
       " 'difference',\n",
       " 'gene_head',\n",
       " 'regression_root',\n",
       " 'ad_root',\n",
       " 'chengdu',\n",
       " 'promote',\n",
       " 'mutation',\n",
       " 'study',\n",
       " 'control',\n",
       " 'tree_dists',\n",
       " 'group_root',\n",
       " 'exist',\n",
       " 'association_root',\n",
       " 'disequilibrium',\n",
       " 'VERB',\n",
       " 'severity',\n",
       " 'NOUN',\n",
       " 'in',\n",
       " 'expression',\n",
       " 'variant',\n",
       " 'case',\n",
       " 'hypotension',\n",
       " 'korean',\n",
       " 'level_root',\n",
       " 'alpha_2-macroglobulin',\n",
       " 'ADJ',\n",
       " 'allele_root',\n",
       " 'datum',\n",
       " 'g',\n",
       " 'smoking',\n",
       " 'apoe4',\n",
       " 'occurrence',\n",
       " 'subject',\n",
       " 'concentration_root',\n",
       " 'ace_root',\n",
       " 'increase_root',\n",
       " 'obese',\n",
       " 'a2_root',\n",
       " 'abcb6',\n",
       " 'conclusion',\n",
       " 'NUM',\n",
       " \"alzheimer's_disease\",\n",
       " 'abca1',\n",
       " 'disease_fork_len',\n",
       " 'finding',\n",
       " 'at1r',\n",
       " 'locus',\n",
       " 'infection',\n",
       " 'variability',\n",
       " 'treatment',\n",
       " 'patient',\n",
       " 'initiation',\n",
       " 'common_word',\n",
       " 'investigate',\n",
       " 'sensitivity',\n",
       " 'lack',\n",
       " 'gene_fork_len',\n",
       " 'form',\n",
       " 'adult',\n",
       " 'snp',\n",
       " 'gracile_syndrome',\n",
       " 'ms',\n",
       " 'frequency',\n",
       " 'number',\n",
       " 'ADV',\n",
       " 'marker',\n",
       " 'PRON',\n",
       " 'disease_root',\n",
       " 'word_dists',\n",
       " 'eno',\n",
       " 'gene_root',\n",
       " 'type_2_diabete',\n",
       " 't_root',\n",
       " 'consequence',\n",
       " 'permissiveness',\n",
       " 'individual',\n",
       " 'a2m.',\n",
       " 'who',\n",
       " 'homozygosity',\n",
       " 'work',\n",
       " 'SYM',\n",
       " 'noun_negated',\n",
       " 'risk',\n",
       " 'vivo',\n",
       " 'rate',\n",
       " 'role',\n",
       " 'CCONJ',\n",
       " 'reflect',\n",
       " 'PART',\n",
       " 'region',\n",
       " 'profile',\n",
       " 'common_depth',\n",
       " 'haplotype',\n",
       " 'SPACE',\n",
       " 'ace_head',\n",
       " 'evidence',\n",
       " 'proportion',\n",
       " 'ireland',\n",
       " 'progression',\n",
       " 'resistance',\n",
       " 'reduce',\n",
       " 'development_root',\n",
       " 'cholelithiasis',\n",
       " 'combination',\n",
       " 'tnf',\n",
       " 'syndrome',\n",
       " 'increase_head',\n",
       " 'cad',\n",
       " 'relate',\n",
       " 'subset',\n",
       " 'strengthen',\n",
       " 'feature',\n",
       " 'history',\n",
       " 'population',\n",
       " 'factor_head',\n",
       " 't_head',\n",
       " 'interaction',\n",
       " 'nature',\n",
       " 'm_root',\n",
       " 'group_head',\n",
       " '-PRON-',\n",
       " 'abcb1',\n",
       " 'obesity_root',\n",
       " 'arg',\n",
       " 'variation_root',\n",
       " 'during',\n",
       " 'change',\n",
       " 'presence',\n",
       " 'X',\n",
       " 'myocardial_infarction',\n",
       " 'regression_head',\n",
       " 'chd',\n",
       " 'prevalence',\n",
       " 'argue',\n",
       " 'analysis',\n",
       " 'before',\n",
       " 'factor_root',\n",
       " 'INTJ',\n",
       " 'm_head',\n",
       " 'within',\n",
       " 'guangzhou_head',\n",
       " 'reveal',\n",
       " 'guangzhou_root',\n",
       " 'dysfunction',\n",
       " 'contribute',\n",
       " 'ID',\n",
       " 'than',\n",
       " 'involve',\n",
       " 'kidney',\n",
       " 'diabetic_nephropathy',\n",
       " 'chymase',\n",
       " 'against',\n",
       " 'have',\n",
       " 'appear',\n",
       " 'warrant',\n",
       " 'breast_cancer',\n",
       " 'merit',\n",
       " 'play',\n",
       " 'need',\n",
       " 'due',\n",
       " 'development_head',\n",
       " '677tt',\n",
       " 'report',\n",
       " 'examine',\n",
       " 'polymorphism_head',\n",
       " 'correlate',\n",
       " 'aortic',\n",
       " 'disease_head',\n",
       " 'idiopathic_dilated_cardiomyopathy',\n",
       " 'association_head',\n",
       " 'nos3-cc',\n",
       " 'variation_head',\n",
       " 'agt',\n",
       " 'concentration_head',\n",
       " 'provide',\n",
       " 'give',\n",
       " 'obesity_head',\n",
       " 'at',\n",
       " 'conclude',\n",
       " 'find',\n",
       " 'on',\n",
       " 'represent',\n",
       " 'modify',\n",
       " 'associate',\n",
       " 'level_head',\n",
       " 'influence',\n",
       " 'd',\n",
       " 'through',\n",
       " 'genotype_head',\n",
       " 'alpha-308',\n",
       " 'show',\n",
       " 'lrp_head',\n",
       " 'indicate',\n",
       " 'between',\n",
       " 'as',\n",
       " 'for',\n",
       " 'support',\n",
       " 'allele_head',\n",
       " 'observe',\n",
       " 'with',\n",
       " 'be',\n",
       " 'confirm',\n",
       " 'use',\n",
       " 'ad_head',\n",
       " 'affect',\n",
       " 'to',\n",
       " 'a2_head',\n",
       " 'of',\n",
       " 'suggest']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coef = clf.coef_[0]\n",
    "coef = [(c, n) for c, n in zip(coef, range(len(coef)))]\n",
    "[train_features.columns[x[1]] for x in sorted(coef, key=lambda x: abs(x[0]))[::-1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['PROPN',\n",
       " 'PUNCT',\n",
       " 'formation',\n",
       " 'lrp_root',\n",
       " 'despite',\n",
       " 'polymorphism_root',\n",
       " 'chengdu',\n",
       " 'homozygote',\n",
       " 'difference',\n",
       " 'promote',\n",
       " 'mutation',\n",
       " 'shared_negation',\n",
       " 'study',\n",
       " 'group_root',\n",
       " 'control',\n",
       " 'disequilibrium',\n",
       " 'severity',\n",
       " 'ad_root',\n",
       " 'enos',\n",
       " 'genotype_root',\n",
       " 'case',\n",
       " 'hypotension',\n",
       " 'variant',\n",
       " 'datum',\n",
       " 'DET',\n",
       " 'from',\n",
       " 'contribution',\n",
       " 'korean',\n",
       " 'alpha_2-macroglobulin',\n",
       " 'ace_root',\n",
       " 'by',\n",
       " 'result',\n",
       " 'concentration_root',\n",
       " 'level_root',\n",
       " 'in',\n",
       " 'expression',\n",
       " 'subject',\n",
       " 'occurrence',\n",
       " 'ADV',\n",
       " 'ADP',\n",
       " 'NUM',\n",
       " 'abcb6',\n",
       " \"alzheimer's_disease\",\n",
       " 'exist',\n",
       " 'association_root',\n",
       " 'g',\n",
       " 'increase_root',\n",
       " 'gene_head',\n",
       " 'regression_root',\n",
       " 'allele_root',\n",
       " 'conclusion',\n",
       " 'treatment',\n",
       " 'at1r',\n",
       " 'locus',\n",
       " 'NOUN',\n",
       " 'finding',\n",
       " 'variability',\n",
       " 'lack',\n",
       " 'abca1',\n",
       " 'number',\n",
       " 'any_negation',\n",
       " 'infection',\n",
       " 'VERB',\n",
       " 'tree_dists',\n",
       " 'smoking',\n",
       " 'apoe4',\n",
       " 'sensitivity',\n",
       " 'initiation',\n",
       " 'investigate',\n",
       " 'gracile_syndrome',\n",
       " 'disease_root',\n",
       " 'form',\n",
       " 'frequency',\n",
       " 'patient',\n",
       " 'marker',\n",
       " 'who',\n",
       " 'ms',\n",
       " 'snp',\n",
       " 'consequence',\n",
       " 'eno',\n",
       " 'gene_root',\n",
       " 'adult',\n",
       " 'type_2_diabete',\n",
       " 'obese',\n",
       " 'a2_root',\n",
       " 'a2m.',\n",
       " 'ireland',\n",
       " 'role',\n",
       " 'PRON',\n",
       " 'individual',\n",
       " 'ADJ',\n",
       " 't_root',\n",
       " 'work',\n",
       " 'region',\n",
       " 'proportion',\n",
       " 'm_root',\n",
       " 'disease_fork_len',\n",
       " 'ace_head',\n",
       " 'evidence',\n",
       " 'permissiveness',\n",
       " 'increase_head',\n",
       " 'cad',\n",
       " 'reflect',\n",
       " 'cholelithiasis',\n",
       " 'profile',\n",
       " 'reduce',\n",
       " 'development_root',\n",
       " 'tnf',\n",
       " 'risk',\n",
       " 'gene_fork_len',\n",
       " 'resistance',\n",
       " 'rate',\n",
       " 'noun_negated',\n",
       " 'homozygosity',\n",
       " 'SYM',\n",
       " 'history',\n",
       " 'population',\n",
       " 'vivo',\n",
       " 'obesity_root',\n",
       " 'variation_root',\n",
       " 't_head',\n",
       " 'interaction',\n",
       " 'nature',\n",
       " 'syndrome',\n",
       " 'word_dists',\n",
       " 'PART',\n",
       " 'group_head',\n",
       " '-PRON-',\n",
       " 'combination',\n",
       " 'abcb1',\n",
       " 'chd',\n",
       " 'myocardial_infarction',\n",
       " 'factor_head',\n",
       " 'during',\n",
       " 'change',\n",
       " 'guangzhou_head',\n",
       " 'SPACE',\n",
       " 'progression',\n",
       " 'common_depth',\n",
       " 'before',\n",
       " 'factor_root',\n",
       " 'regression_head',\n",
       " 'argue',\n",
       " 'analysis',\n",
       " 'haplotype',\n",
       " 'arg',\n",
       " 'dysfunction',\n",
       " 'prevalence',\n",
       " 'm_head',\n",
       " 'relate',\n",
       " 'subset',\n",
       " 'INTJ',\n",
       " 'strengthen',\n",
       " 'feature',\n",
       " 'guangzhou_root',\n",
       " 'within',\n",
       " 'presence',\n",
       " 'reveal',\n",
       " 'CCONJ',\n",
       " 'contribute',\n",
       " 'X',\n",
       " 'than',\n",
       " 'involve',\n",
       " 'kidney',\n",
       " 'diabetic_nephropathy',\n",
       " 'chymase',\n",
       " 'against',\n",
       " 'have',\n",
       " 'appear',\n",
       " 'warrant',\n",
       " 'breast_cancer',\n",
       " 'merit',\n",
       " 'play',\n",
       " 'need',\n",
       " 'due',\n",
       " 'development_head',\n",
       " '677tt',\n",
       " 'report',\n",
       " 'examine',\n",
       " 'polymorphism_head',\n",
       " 'correlate',\n",
       " 'aortic',\n",
       " 'disease_head',\n",
       " 'idiopathic_dilated_cardiomyopathy',\n",
       " 'association_head',\n",
       " 'nos3-cc',\n",
       " 'variation_head',\n",
       " 'agt',\n",
       " 'concentration_head',\n",
       " 'provide',\n",
       " 'give',\n",
       " 'obesity_head',\n",
       " 'at',\n",
       " 'conclude',\n",
       " 'find',\n",
       " 'on',\n",
       " 'represent',\n",
       " 'modify',\n",
       " 'associate',\n",
       " 'level_head',\n",
       " 'influence',\n",
       " 'd',\n",
       " 'through',\n",
       " 'genotype_head',\n",
       " 'alpha-308',\n",
       " 'show',\n",
       " 'lrp_head',\n",
       " 'indicate',\n",
       " 'between',\n",
       " 'as',\n",
       " 'for',\n",
       " 'support',\n",
       " 'allele_head',\n",
       " 'observe',\n",
       " 'with',\n",
       " 'be',\n",
       " 'confirm',\n",
       " 'use',\n",
       " 'ad_head',\n",
       " 'affect',\n",
       " 'to',\n",
       " 'a2_head',\n",
       " 'of',\n",
       " 'suggest']"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_features = pd.read_csv('../corpus/features/left_features_train.csv',\n",
    "                             usecols= lambda s: s !='common_word')\n",
    "coef = clf.coef_[0]\n",
    "coef = [(c, n) for c, n in zip(coef, range(len(coef)))]\n",
    "[train_features.columns[x[1]] for x in sorted(coef, key=lambda x: abs(x[0]))[::-1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.8409785932721713\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(0.8493624683606741, 0.8409785932721713, 0.8324466065463437, None)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf = LogisticRegression(solver='liblinear', multi_class='auto').fit(train_features, train_labels)\n",
    "print(sum([p == a for p, a in zip(clf.predict(test_features),test_labels)]) / len(test_labels))\n",
    "precision_recall_fscore_support(test_labels, clf.predict(test_features), average='weighted')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.8226299694189603\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(0.8387502074295332, 0.8226299694189603, 0.8091385256034874, None)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svm_clf = SVC(gamma='scale').fit(train_features, train_labels)\n",
    "print(sum([p == a for p, a in zip(svm_clf.predict(test_features),test_labels)]) / len(test_labels))\n",
    "precision_recall_fscore_support(test_labels, svm_clf.predict(test_features), average='weighted')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_labels = pd.read_csv('../dataset/final_train.csv',\n",
    "                          usecols=[1])\n",
    "train_labels = list(train_labels['Association'])\n",
    "test_labels = pd.read_csv('../dataset/final_test.csv',\n",
    "                          usecols=[1])\n",
    "test_labels = list(test_labels['Association'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7063291139240506\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(0.704817158931083, 0.7063291139240506, 0.6765160457131143, None)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf = LogisticRegression(solver='lbfgs', multi_class='auto', max_iter=1000).fit(train_features, train_labels)\n",
    "print(sum([p == a for p, a in zip(clf.predict(test_features),test_labels)]) / len(test_labels))\n",
    "precision_recall_fscore_support(test_labels, clf.predict(test_features), average='weighted')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['NOUN',\n",
       " 'common_word',\n",
       " 'ADP',\n",
       " 'PUNCT',\n",
       " 'by',\n",
       " 'result',\n",
       " 'PROPN',\n",
       " 'shared_negation',\n",
       " 'VERB',\n",
       " 'tree_dists',\n",
       " 'DET',\n",
       " 'from',\n",
       " 'contribution',\n",
       " 'lrp_root',\n",
       " 'common_depth',\n",
       " 'ADJ',\n",
       " 'gene_head',\n",
       " 'regression_root',\n",
       " 'despite',\n",
       " 'polymorphism_root',\n",
       " 'enos',\n",
       " 'genotype_root',\n",
       " 'formation',\n",
       " 'word_dists',\n",
       " 'ad_root',\n",
       " 'chengdu',\n",
       " 'homozygote',\n",
       " 'difference',\n",
       " 'obese',\n",
       " 'a2_root',\n",
       " 'promote',\n",
       " 'mutation',\n",
       " 'gene_fork_len',\n",
       " 'study',\n",
       " 'group_head',\n",
       " '-PRON-',\n",
       " 'disease_fork_len',\n",
       " 'control',\n",
       " 'reduce',\n",
       " 'development_root',\n",
       " 'variant',\n",
       " 'group_root',\n",
       " 'allele_root',\n",
       " 'in',\n",
       " 'expression',\n",
       " 'korean',\n",
       " 'exist',\n",
       " 'association_root',\n",
       " 'noun_negated',\n",
       " 'level_root',\n",
       " 'disequilibrium',\n",
       " 'case',\n",
       " 'severity',\n",
       " 'alpha_2-macroglobulin',\n",
       " 'hypotension',\n",
       " 'any_negation',\n",
       " 'conclusion',\n",
       " 'patient',\n",
       " 'g',\n",
       " 'increase_root',\n",
       " 'smoking',\n",
       " 'apoe4',\n",
       " 'abcb6',\n",
       " 'during',\n",
       " 'change',\n",
       " 'occurrence',\n",
       " 'concentration_root',\n",
       " 'ace_root',\n",
       " 'frequency',\n",
       " 'form',\n",
       " 'ADV',\n",
       " 'before',\n",
       " 'factor_root',\n",
       " 'subject',\n",
       " 'initiation',\n",
       " 'ms',\n",
       " 'variability',\n",
       " 'sensitivity',\n",
       " 'CCONJ',\n",
       " 'investigate',\n",
       " 'PART',\n",
       " 'finding',\n",
       " 'history',\n",
       " 'population',\n",
       " 'snp',\n",
       " 'at1r',\n",
       " 'locus',\n",
       " 'PRON',\n",
       " 'infection',\n",
       " 'haplotype',\n",
       " 'individual',\n",
       " 'disease_root',\n",
       " 'abca1',\n",
       " 'adult',\n",
       " 't_root',\n",
       " 'SYM',\n",
       " \"alzheimer's_disease\",\n",
       " 'marker',\n",
       " 'number',\n",
       " 'permissiveness',\n",
       " 'homozygosity',\n",
       " 'vivo',\n",
       " 'a2m.',\n",
       " 'progression',\n",
       " 'SPACE',\n",
       " 'consequence',\n",
       " 'treatment',\n",
       " 'risk',\n",
       " 'reflect',\n",
       " 'type_2_diabete',\n",
       " 'rate',\n",
       " 'work',\n",
       " 'combination',\n",
       " 'lack',\n",
       " 'NUM',\n",
       " 'gracile_syndrome',\n",
       " 'variation_root',\n",
       " 'argue',\n",
       " 'analysis',\n",
       " 'who',\n",
       " 'proportion',\n",
       " 'profile',\n",
       " 'role',\n",
       " 'region',\n",
       " 'cholelithiasis',\n",
       " 'abcb1',\n",
       " 'resistance',\n",
       " 'ace_head',\n",
       " 'evidence',\n",
       " 'strengthen',\n",
       " 'feature',\n",
       " 'eno',\n",
       " 'gene_root',\n",
       " 'syndrome',\n",
       " 'tnf',\n",
       " 'factor_head',\n",
       " 'X',\n",
       " 'chd',\n",
       " 'prevalence',\n",
       " 'guangzhou_root',\n",
       " 'dysfunction',\n",
       " 'arg',\n",
       " 't_head',\n",
       " 'interaction',\n",
       " 'nature',\n",
       " 'datum',\n",
       " 'relate',\n",
       " 'subset',\n",
       " 'm_root',\n",
       " 'guangzhou_head',\n",
       " 'regression_head',\n",
       " 'presence',\n",
       " 'm_head',\n",
       " 'INTJ',\n",
       " 'contribute',\n",
       " 'myocardial_infarction',\n",
       " 'increase_head',\n",
       " 'cad',\n",
       " 'within',\n",
       " 'reveal',\n",
       " 'obesity_root',\n",
       " 'ireland',\n",
       " 'ID',\n",
       " 'than',\n",
       " 'involve',\n",
       " 'kidney',\n",
       " 'diabetic_nephropathy',\n",
       " 'chymase',\n",
       " 'against',\n",
       " 'have',\n",
       " 'appear',\n",
       " 'warrant',\n",
       " 'breast_cancer',\n",
       " 'merit',\n",
       " 'play',\n",
       " 'need',\n",
       " 'due',\n",
       " 'development_head',\n",
       " '677tt',\n",
       " 'report',\n",
       " 'examine',\n",
       " 'polymorphism_head',\n",
       " 'correlate',\n",
       " 'aortic',\n",
       " 'disease_head',\n",
       " 'idiopathic_dilated_cardiomyopathy',\n",
       " 'association_head',\n",
       " 'nos3-cc',\n",
       " 'variation_head',\n",
       " 'agt',\n",
       " 'concentration_head',\n",
       " 'provide',\n",
       " 'give',\n",
       " 'obesity_head',\n",
       " 'at',\n",
       " 'conclude',\n",
       " 'find',\n",
       " 'on',\n",
       " 'represent',\n",
       " 'modify',\n",
       " 'associate',\n",
       " 'level_head',\n",
       " 'influence',\n",
       " 'd',\n",
       " 'through',\n",
       " 'genotype_head',\n",
       " 'alpha-308',\n",
       " 'show',\n",
       " 'lrp_head',\n",
       " 'indicate',\n",
       " 'between',\n",
       " 'as',\n",
       " 'for',\n",
       " 'support',\n",
       " 'allele_head',\n",
       " 'observe',\n",
       " 'with',\n",
       " 'be',\n",
       " 'confirm',\n",
       " 'use',\n",
       " 'ad_head',\n",
       " 'affect',\n",
       " 'to',\n",
       " 'a2_head',\n",
       " 'of',\n",
       " 'suggest']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coef = clf.coef_[0]\n",
    "coef = [(c, n) for c, n in zip(coef, range(len(coef)))]\n",
    "[train_features.columns[x[1]] for x in sorted(coef, key=lambda x: abs(x[0]))[::-1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.533457249070632\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(0.5279145773449703, 0.533457249070632, 0.52274144662177, None)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf = LogisticRegression(solver='lbfgs', multi_class='auto', max_iter=1000).fit(train_features, train_labels)\n",
    "print(sum([p == a for p, a in zip(clf.predict(test_features),test_labels)]) / len(test_labels))\n",
    "precision_recall_fscore_support(test_labels, clf.predict(test_features), average='weighted')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'Y': 1192, 'N': 658, 'F': 384})\n",
      "Counter({'Y': 212, 'N': 115, 'F': 68})\n",
      "Counter({'Y': 274, 'N': 89, 'F': 32})\n"
     ]
    }
   ],
   "source": [
    "print(Counter(train_labels))\n",
    "print(Counter(test_labels))\n",
    "print(Counter(clf.predict(test_features)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.6936708860759494\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(0.6979201503841768, 0.6936708860759494, 0.6709524378459842, None)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svm_clf = SVC(gamma='scale').fit(train_features, train_labels)\n",
    "print(sum([p == a for p, a in zip(svm_clf.predict(test_features),test_labels)]) / len(test_labels))\n",
    "precision_recall_fscore_support(test_labels, svm_clf.predict(test_features), average='weighted')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split(a, n):\n",
    "    k, m = divmod(len(a), n)\n",
    "    return (a[i * k + min(i, m):(i + 1) * k + min(i + 1, m)] for i in range(n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7088607594936709"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_splits = split(train_features, 5)\n",
    "train_labels_splits = split(train_labels, 5)\n",
    "clfs = []\n",
    "for data, labels in zip(train_splits, train_labels_splits):\n",
    "    eclf = LogisticRegression(solver='lbfgs', multi_class='auto', max_iter=1000).fit(data, labels)\n",
    "    clfs.append(eclf)\n",
    "preds = [c.predict(test_features) for c in clfs]\n",
    "pred = [max(set(x), key=x.count) for x in zip(*preds)]\n",
    "sum([p == a for p, a in zip(pred,test_labels)]) / len(test_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n",
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.8481012658227848\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jsilva/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.\n",
      "  \"of iterations.\", ConvergenceWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(array([0.63333333, 0.86575342]),\n",
       " array([0.27941176, 0.96636086]),\n",
       " array([0.3877551, 0.9132948]),\n",
       " array([ 68, 327]))"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf = LogisticRegressionCV(cv=5, solver='lbfgs', multi_class='auto').fit(train_features, train_labels)\n",
    "print(sum([p == a for p, a in zip(clf.predict(test_features),test_labels)]) / len(test_labels))\n",
    "precision_recall_fscore_support(test_labels, clf.predict(test_features), average=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
